python/tests/test_extend_edges.py

```diff
@@ -0,0 +1,672 @@
+import msprime
+import numpy as np
+import pytest
+
+import _tskit
+import tests.test_wright_fisher as wf
+import tskit
+from tests import tsutil
+from tests.test_highlevel import get_example_tree_sequences
+
+# ↑ See https://github.com/tskit-dev/tskit/issues/1804 for when
+# we can remove this.
+
+
+def extend_edges(ts, max_iter=10):
+    tables = ts.dump_tables()
+    mutations = tables.mutations.copy()
+    tables.mutations.clear()
+
+    last_num_edges = ts.num_edges
+    for _ in range(max_iter):
+        for forwards in [True, False]:
+            edges = _extend(ts, forwards=forwards)
+            tables.edges.replace_with(edges)
+            tables.build_index()
+            ts = tables.tree_sequence()
+        if ts.num_edges == last_num_edges:
+            break
+        else:
+            last_num_edges = ts.num_edges
+
+    tables = ts.dump_tables()
+    mutations = _slide_mutation_nodes_up(ts, mutations)
+    tables.mutations.replace_with(mutations)
+    tables.build_index()
+    ts = tables.tree_sequence()
+
+    return ts
+
+
+def _slide_mutation_nodes_up(ts, mutations):
+    # adjusts mutations' nodes to place each mutation on the correct edge given
+    # their time; requires mutation times be nonmissing and the mutation times
+    # be >= their nodes' times.
+
+    assert np.all(~tskit.is_unknown_time(mutations.time)), "times must be known"
+    parent = np.full(ts.num_nodes, tskit.NULL)
+    edges = ts.tables.edges
+    new_nodes = mutations.node.copy()
+
+    next_mut = 0
+    tree_pos = tsutil.TreePosition(ts)
+    valid = tree_pos.next()
+    while valid:
+        left, right = tree_pos.interval
+
+        for j in range(
+            tree_pos.out_range.start, tree_pos.out_range.stop, tskit.FORWARD
+        ):
+            e = tree_pos.out_range.order[j]
+            parent[edges.child[e]] = -1
+
+        for j in range(tree_pos.in_range.start, tree_pos.in_range.stop, tskit.FORWARD):
+            e = tree_pos.in_range.order[j]
+            parent[edges.child[e]] = edges.parent[e]
+
```

Comment:
That was because we prototyped in python first, then moved it to C.

Reply:
Oh I see - okay, really it was because I was copying what we did in the other function.

Reply:
Actually - it's not that simple because the mutations aren't actually on the trees - they're passed in via a separate table (because they wouldn't be legal on that tree sequence). But it can be simpler, you're right.